File: ShadowFrameworkLite20/src/shadow/geometry/functions/SFCurvedTubeFunction.java
package shadow.geometry.functions;

import shadow.geometry.SFCurve;
import shadow.math.SFVertex3f;

public class SFCurvedTubeFunction extends SFUnoptimizedSurfaceFunction {
	
	private SFCurve centralCurve;
	private SFCurve rayCurve;
	
	private SFVertex3f Ccv=new SFVertex3f();
	private SFVertex3f Vec1v=new SFVertex3f();
	private SFVertex3f Vec2v=new SFVertex3f();
	private SFVertex3f DVec1v=new SFVertex3f();
	private SFVertex3f DVec2v=new SFVertex3f();
	private SFVertex3f dCcdv=new SFVertex3f();
	
	public SFCurvedTubeFunction() {
		super();
	}
	
	public SFCurvedTubeFunction(SFCurve centralCurve,
			SFCurve rayCurve) {
		super();
		this.centralCurve = centralCurve;
		this.rayCurve = rayCurve;
	}
	
	public SFVertex3f getDu() {
		SFVertex3f du=new SFVertex3f(getDXDu(),getDYDu(),getDZDu());
		du.normalize3f();
		return du;
	}

	public SFVertex3f getDv() {
		SFVertex3f dv=new SFVertex3f(getDXDv(),getDYDv(),getDZDv());
		dv.normalize3f();
		return dv;
	}

	public SFVertex3f getPosition() {
		return new SFVertex3f(getX(),getY(),getZ());
	}

	
	@Override
	public void init() {
		//Do nothing
	}
	
	@Override
	public void destroy() {
		//Its correct: if init isn't doing anything, destroy should not do anything
	}

	private float lastV=-1;
	private float cos=0;
	private float sin=0;
	
	
	private void evalAll(float v){
		if(v!=lastV){
			centralCurve.getVertex(v, Ccv);
			centralCurve.getDevDt(v, dCcdv);
			rayCurve.getVertex(v, Vec1v);
			Vec1v.subtract3f(Ccv);
			Vec2v.set3f(dCcdv.cross(Vec1v));
			Vec2v.mult((float)(Math.sqrt(Vec1v.dot3f(Vec1v)/Vec2v.dot3f(Vec2v))));

			SFVertex3f dCcdv2=new SFVertex3f();
			centralCurve.getDev2Dt(v, dCcdv2);
			rayCurve.getDevDt(v, DVec1v);
			DVec1v.subtract3f(dCcdv);
			//DVec1v.normalize3f();
			DVec2v.set3f(dCcdv.cross(DVec1v));
			DVec2v.add3f(dCcdv2.cross(Vec1v));
			//DVec2v.normalize3f();
			
			lastV=v;
		}
	}
	
	public float getX(float u,float v) {
		evalAll(v);
		cos=(float)(Math.cos(2*Math.PI*u));
		sin=(float)(Math.sin(2*Math.PI*u));
		return Ccv.getX()+cos*Vec1v.getX()+sin*Vec2v.getX();
	}
	
	public float getY(float u,float v) {
		return Ccv.getY()+cos*Vec1v.getY()+sin*Vec2v.getY();
	}
	
	public float getZ(float u,float v) {
		return Ccv.getZ()+cos*Vec1v.getZ()+sin*Vec2v.getZ();
	}
	
	
	public float getX() {
		return Ccv.getX()+cos*Vec1v.getX()+sin*Vec2v.getX();
	}
	
	public float getY() {
		return Ccv.getY()+cos*Vec1v.getY()+sin*Vec2v.getY();
	}
	
	public float getZ() {
		return Ccv.getZ()+cos*Vec1v.getZ()+sin*Vec2v.getZ();
	}
	
	public float getDXDv() {
		return dCcdv.getX()+cos*DVec1v.getX()+sin*DVec2v.getX();
	}
	
	public float getDYDv() {
		return dCcdv.getY()+cos*DVec1v.getY()+sin*DVec2v.getY();
	}
	
	public float getDZDv() {
		return dCcdv.getZ()+cos*DVec1v.getZ()+sin*DVec2v.getZ();
	}
	
	public float getDXDu() {
		return -sin*Vec1v.getX()+cos*Vec2v.getX();
	}
	
	public float getDYDu() {
		return -sin*Vec1v.getY()+cos*Vec2v.getY();
	}
	
	public float getDZDu() {
		return -sin*Vec1v.getZ()+cos*Vec2v.getZ();
	}

	public SFCurve getCentralCurve() {
		return centralCurve;
	}

	public void setCentralCurve(SFCurve centralCurve) {
		this.centralCurve = centralCurve;
	}

	public SFCurve getRayCurve() {
		return rayCurve;
	}

	public void setRayCurve(SFCurve rayCurve) {
		this.rayCurve = rayCurve;
	}

	

}

File: trunk/ShadowFrameworkJS/src/shadow/geometry/functions/SFCurvedTubeFunction.js


function SFCurvedTubeFunction(){
	
	this.Ccv=new SFVertex3f();
	this.Vec1v=new SFVertex3f();
	this.Vec2v=new SFVertex3f();
	this.DVec1v=new SFVertex3f();
	this.DVec2v=new SFVertex3f();
	this.dCcdv=new SFVertex3f();
	this.lastV=-1;
	this.cos=0;
	this.sin=0;
	
}

inherit(SFCurvedTubeFunction,SFUnoptimizedSurfaceFunction);

/*		
SFCurvedTubeFunction.prototype["getDu"]=function(f){
	var du=new SFVertex3f(this.getDXDu(),this.getDYDu(),this.getDZDu());
	du.normalize3f();
	return du;
};

SFCurvedTubeFunction.prototype["getDv"]=function(f){
	var dv=new SFVertex3f(this.getDXDv(),this.getDYDv(),this.getDZDv());
	dv.normalize3f();
	return dv;
};
*/

SFCurvedTubeFunction.prototype["init"]=function(){
	//DO nothing
	;
};


SFCurvedTubeFunction.prototype["destroy"]=function(){
	//DO nothing
	;
};

SFCurvedTubeFunction.prototype["destroy"]=function(){
	//DO nothings
	;
};

SFCurvedTubeFunction.prototype["evalAll"]=function(v){
	if(v!=this.lastV){

		this.centralCurve.getVertex(v, this.Ccv);

		this.centralCurve.getDevDt(v, this.dCcdv);
		
		this.rayCurve.getVertex(v, this.Vec1v);
		
		this.Vec1v.subtract3f(this.Ccv);
		this.Vec2v.set3f(this.dCcdv.cross(this.Vec1v));
		this.Vec2v.mult((Math.sqrt(this.Vec1v.dot3f(this.Vec1v)/this.Vec2v.dot3f(this.Vec2v))));
		
		this.dCcdv2=new SFVertex3f();
		this.centralCurve.getDev2Dt(v, this.dCcdv2);
		this.rayCurve.getDevDt(v, this.DVec1v);
		this.DVec1v.subtract3f(this.dCcdv);
		
		//DVec1v.normalize3f();
		this.DVec2v.set3f(this.dCcdv.cross(this.DVec1v));
		
		this.DVec2v.add3fv(this.dCcdv2.cross(this.Vec1v));
		
		//DVec2v.normalize3f();
		
		
		this.lastV=v;
	}
};

SFCurvedTubeFunction.prototype["getX"]=function(u,v){
	this.evalAll(v);
	this.cos=(Math.cos(2*Math_PI*u));
	this.sin=(Math.sin(2*Math_PI*u));
	
	return this.Ccv.getX()+this.cos*this.Vec1v.getX()+this.sin*this.Vec2v.getX();
};

SFCurvedTubeFunction.prototype["getY"]=function(u,v){
	return this.Ccv.getY()+this.cos*this.Vec1v.getY()+this.sin*this.Vec2v.getY();
};

SFCurvedTubeFunction.prototype["getZ"]=function(u,v){
	return this.Ccv.getZ()+this.cos*this.Vec1v.getZ()+this.sin*this.Vec2v.getZ();
};


SFCurvedTubeFunction.prototype["getDXDv"]=function(u,v){
	return this.dCcdv.getX()+this.cos*this.DVec1v.getX()+this.sin*this.DVec2v.getX();
};

SFCurvedTubeFunction.prototype["getDYDv"]=function(u,v){
	return this.dCcdv.getY()+this.cos*this.DVec1v.getY()+this.sin*this.DVec2v.getY();
};

SFCurvedTubeFunction.prototype["getDZDv"]=function(u,v){
	return this.dCcdv.getZ()+this.cos*this.DVec1v.getZ()+this.sin*this.DVec2v.getZ();
};


SFCurvedTubeFunction.prototype["getDXDu"]=function(u,v){
	return -this.sin*this.Vec1v.getX()+this.cos*this.Vec2v.getX();
};

SFCurvedTubeFunction.prototype["getDYDu"]=function(u,v){
	return -this.sin*this.Vec1v.getY()+this.cos*this.Vec2v.getY();
};

SFCurvedTubeFunction.prototype["getDZDu"]=function(u,v){
	return -this.sin*this.Vec1v.getZ()+this.cos*this.Vec2v.getZ();
};


SFCurvedTubeFunction.prototype["getCentralCurve"]=function(){
	return this.centralCurve;
};

SFCurvedTubeFunction.prototype["setCentralCurve"]=function(centralCurve){
	this.centralCurve=centralCurve;
};

SFCurvedTubeFunction.prototype["getRayCurve"]=function(){
	return this.rayCurve;
};

SFCurvedTubeFunction.prototype["setRayCurve"]=function(rayCurve){
	this.rayCurve=rayCurve;
};
